### PR TITLE
Enable CPP/CUDAExtension with py_limited_api for python agnosticism

### PR DESCRIPTION
**Pull request by janeyx99**
Getting tested with ao, but now there is a real test i added.

## What does this PR do?

We want to allow custom PyTorch extensions to be able to build one wheel for multiple Python versions, in other words, achieve python agnosticism. It turns out that there is such a way that setuptools/Python provides already! Namely, if the user promises to use only the Python limited API in their extension, they can pass in `py_limited_api` to their Extension class and to the bdist_wheel command (with a min python version) in order to build 1 wheel that will suffice across multiple Python versions.

Sounds lovely! Why don't people do that already with PyTorch? Well 2 things. This workflow is hardly documented (even searching for python agnostic specifically does not reveal many answers) so I'd expect that people simply don't know about it. But even if they did, _PyTorch_ custom Extensions would still not work because we always link torch_python, which does not abide by py_limited_api rules.

So this is where this PR comes in! We respect when the user specifies py_limited_api and skip linking torch_python under that condition, allowing users to enroll in the provided functionality I just described.

## How do I know this PR works?

I manually tested my silly little ultra_norm locally (with `import python_agnostic`) and wrote a test case for the extension showing that
- torch_python doesn't show up in the ldd tree
- no Py- symbols show up
It may be a little confusing that our test case is actually python-free (more clean than python-agnostic) but it is sufficient (and not necessary) towards showing that this change works.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #138088

### CODE REVIEW

File: test/cpp_extensions/python_agnostic_extension/python_agnostic/ops.py
Comment (janeyx99) [inline test/cpp_extensions/python_agnostic_extension/python_agnostic/ops.py:1]: This is copied from ao...we could simplify it a lot more

File: test/cpp_extensions/python_agnostic_extension/python_agnostic/csrc/tensor_core_tiled_layout.cu
Comment (janeyx99) [inline test/cpp_extensions/python_agnostic_extension/python_agnostic/csrc/tensor_core_tiled_layout.cu:1]: also copied from ao as an example

Comment (ezyang) [reply test/cpp_extensions/python_agnostic_extension/python_agnostic/csrc/tensor_core_tiled_layout.cu:1]: This is probably too complicated for what you actually need as a unit test. And it's not clear we are getting *that* much value out of a full size example.

File: test/test_cpp_extensions_aot.py [hunk 1]
```diff
@@ -164,6 +168,48 @@ def test_cuda_dlink_libs(self):
...
+        #   b. The produced shared library does not have libtorch_python.so as a
+        # dependency from the output of "ldd _C.so"
+        #   c. The .so does not need any python related symbols. We approximate
+        # this by running "nm -u _C.so" and grepping that nothing starts with "Py"
+
+        dist_root = os.path.join("cpp_extensions", "python_agnostic_extension", "dist")
+        matches = list(Path(dist_root).glob("*.whl"))
+        self.assertEqual(len(matches), 1, msg=str(matches))
+        whl_file = matches[0]
+        self.assertRegex(str(whl_file), r".*python_agnostic-0\.0-cp39-abi3-.*\.whl")
```
Comment (ezyang) [inline test/test_cpp_extensions_aot.py:189]: I'm confused: why do you require minimum CPython version 3.9? You're not using /any/ CPython APIs right?

**Review by ezyang — APPROVED** [review]
see comments

File: test/cpp_extensions/python_agnostic_extension/python_agnostic/csrc/tensor_core_tiled_layout.cu [hunk 2]
```diff
@@ -0,0 +1,316 @@
+#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 800  // at least Ampere
```
Comment (albanD) [inline test/cpp_extensions/python_agnostic_extension/python_agnostic/csrc/tensor_core_tiled_layout.cu:1]: +1 from the comment above that we can shrink this.
We just need something minimal that depends on cuda and aten right?

File: test/cpp_extensions/python_agnostic_extension/setup.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+def get_extension():
+    extra_compile_args = {
+        "cxx": [
+            "-O3",
+            "-fdiagnostics-color=always",
+        ],
+        "nvcc": [
+            "-O3",
+            "-t=0",
```
Comment (albanD) [inline test/cpp_extensions/python_agnostic_extension/setup.py:46]: Shouldn't this be the default?

File: test/test_cpp_extensions_aot.py (hunk 1 above)
Comment (albanD) [reply test/test_cpp_extensions_aot.py:189]: Because that's the minimum version we support?

File: torch/utils/cpp_extension.py [hunk 4]
```diff
@@ -994,7 +994,9 @@ def CppExtension(name, sources, *args, **kwargs):
     libraries.append('c10')
     libraries.append('torch')
     libraries.append('torch_cpu')
-    libraries.append('torch_python')
+    if not kwargs.get('py_limited_api', False):
```
Comment (albanD) [inline torch/utils/cpp_extension.py:997]: From offline discussion, this flag is not actually related to the limited API. So let's keep it to "no torch python" or something along those line.

File: test/test_cpp_extensions_aot.py (hunk 1 above)
Comment (janeyx99) [reply test/test_cpp_extensions_aot.py:189]: Yea, I think Ed's confusion is of the same essence as what you had asked yesterday, so let me answer here for posterity:

There is indeed a difference between the test case here and the API that is offered, namely:
(1) the custom extension here (or torchao haha) does NOT use any CPython APIs, and is therefore Python-free
but (2) the py-limited-api flag that CPython/Setup/PyPI respects allows for limited CPython use, and only signifies Python-agnosticism.

Our end goal is allowing custom extensions to build 1 python wheel that should work with multiple versions of Python, which, without PyTorch, Python-related-entities already have an API for! To build a wheel that is python agnostic today, one must pass `py_limited_api` to both the underlying [Extension](https://github.com/pypa/setuptools/blob/9692cde009af4651819d18a1e839d3b6e3fcd77d/setuptools/extension.py#L127) class and the [bdist_wheel command](https://github.com/pypa/setuptools/issues/4741). 

This workflow that works for most normal Extensions just doesn't work for PyTorch Extensions because we link torch_python no matter what :') SO the change we're making on our end is not linking torch_python when the flag is set. I will add some documentation making this exclusion explicit to users. (This workflow is terribly documented and there are not even 500 uses on public [GitHub](https://github.com/search?q=%2F%28%3F-i%29%22py_limited_api%22%2F&type=code), so I'm not surprised no one's yelled at us yet.)

After this PR, I do plan on writing a small tutorial showing how someone could do this.

File: torch/utils/cpp_extension.py (hunk 4 above)
Comment (janeyx99) [reply torch/utils/cpp_extension.py:997]: Never mind, I don't agree anymore! :D It is related in that it is a documented/supported thing setuptools offers, so we should deal with this flag no matter what. https://github.com/pypa/setuptools/blob/9692cde009af4651819d18a1e839d3b6e3fcd77d/setuptools/extension.py#L127

We could offer _another_ flag to be explicit about not linking torch_python, but given that the intended `py_limited_api` doesn't work today, I'd prefer documenting that we're fixing that with explicit documentation that torch_python + its offered APIs will not be included. What do you think?

File: test/cpp_extensions/python_agnostic_extension/python_agnostic/csrc/tensor_core_tiled_layout.cu (hunk 2 above)
Comment (janeyx99) [reply test/cpp_extensions/python_agnostic_extension/python_agnostic/csrc/tensor_core_tiled_layout.cu:1]: yeh

File: test/cpp_extensions/python_agnostic_extension/setup.py (hunk 3 above)
Comment (janeyx99) [reply test/cpp_extensions/python_agnostic_extension/setup.py:46]: I actually cannot find a manpage explaining -t=0 (do you know what that is?) for nvcc. In either case, having compiler flags that I don't understand for a unit test is silly, so I'm removing the flags.

Comment (albanD) [reply test/cpp_extensions/python_agnostic_extension/setup.py:46]: https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/#threads-number-t

File: torch/utils/cpp_extension.py [hunk 5]
```diff
@@ -964,6 +964,16 @@ def CppExtension(name, sources, *args, **kwargs):
     constructor. Full list arguments can be found at
     https://setuptools.pypa.io/en/latest/userguide/ext_modules.html#extension-api-reference
 
+    Note!
```
Comment (janeyx99) [inline torch/utils/cpp_extension.py:967]: This _is_ mostly copy/pasta with the CUDAExtension description...I _could_ move it to deduplicate but then we'd have to do our setting __doc__ trick which is less readable in the code....

File: torch/utils/cpp_extension.py [hunk 6]
```diff
@@ -1017,6 +1029,16 @@ def CUDAExtension(name, sources, *args, **kwargs):
...
```
Comment (albanD) [inline torch/utils/cpp_extension.py:1032]: Shouldn't this be `..note:` or something like that?

File: torch/utils/cpp_extension.py [hunk 7]
```diff
@@ -1017,6 +1029,16 @@ def CUDAExtension(name, sources, *args, **kwargs):
...
+    If passing in the ``py_limited_api`` flag signifying that the extension uses
+    only the Python limited API, CUDAExtension will respect the flag and SKIP
+    linking torch_python with your custom extension!!!! This means that your
+    extension CANNOT rely on any APIs provided in torch_python!!!! Notably, your
+    extension cannot call PYBIND11_MODULE, which is commonly used to create a
+    Python module associated with the extension. The APIs in torch_python are
+    blatantly NOT within the Python limited API, which is why we avoid linking
+    torch_python when the ``py_limited_api`` is specified.
```
Comment (albanD) [inline torch/utils/cpp_extension.py:1040]: I think we could make this shorter and to the point? I feel like there are only 2 ideas here:
- PyTorch python API (as provided in libtorch_python) cannot be built with py_limited_api
- When this flag is passed, it is the user responsibility in their library to not use APIs from libtorch_python (in particular pytorch/python bindings) and only use APIs from libtorch (aten objects, operators and registering operations through the dispatcher for access from python).

File: torch/utils/cpp_extension.py (hunk 6 above)
Comment (janeyx99) [reply torch/utils/cpp_extension.py:1032]: I wasn't swayed either way, the cudaextension one had a big chunk of text so I stuck with it. but yea, can change

File: torch/utils/cpp_extension.py (hunk 7 above)
Comment (janeyx99) [reply torch/utils/cpp_extension.py:1040]: oki let me try again!

**Review by albanD — APPROVED** [review]
SGTM !

**Comment by janeyx99** [comment]
@pytorchbot cherry-pick --onto release/2.6 -c fixnewfeature

This fixes py_limited_api usage for torchao to be able to be python agnostic in their releases, as their releases depend on stable torch to build.
